File: src/main/java/ru/trukhmanov/twochairsbackend/service/DeckService.java
package ru.trukhmanov.twochairsbackend.service;

import org.springframework.stereotype.Service;
import ru.trukhmanov.twochairsbackend.dto.game.deck.DeckDto;
import ru.trukhmanov.twochairsbackend.entity.Deck;
import ru.trukhmanov.twochairsbackend.repository.game.DeckRepository;
import ru.trukhmanov.twochairsbackend.repository.UserPurchaseRepository;
import ru.trukhmanov.twochairsbackend.repository.UserRepository;

import java.util.List;
import java.util.Objects;

@Service
public class DeckService {

    private final DeckRepository deckRepository;
    private final UserRepository userRepository;
    private final UserPurchaseRepository purchaseRepository;

    public DeckService(DeckRepository deckRepository,
                       UserRepository userRepository,
                       UserPurchaseRepository purchaseRepository) {
        this.deckRepository = deckRepository;
        this.userRepository = userRepository;
        this.purchaseRepository = purchaseRepository;
    }

    public List<DeckDto> getAccessibleDecks(long userId) {
        boolean premium = userRepository.findById(userId).orElseThrow().isPremium();

        List<Deck> decks = premium
                ? deckRepository.findAccessibleForPremium(userId)
                : deckRepository.findAccessibleForNonPremium(userId);

        return decks.stream()
                .map(d -> toDto(d, false))
                .toList();
    }

    public List<DeckDto> getStoreDecks(long userId) {
        boolean premium = userRepository.findById(userId).orElseThrow().isPremium();
        List<Deck> decks = deckRepository.findStorePaidDecks();

        return decks.stream().map(d -> {
            boolean locked;
            if (premium) locked = false;
            else locked = !purchaseRepository.hasDeck(userId, d.getId());
            return toDto(d, locked);
        }).toList();
    }

    private DeckDto toDto(Deck d, boolean locked) {
        return new DeckDto(
                d.getId(),
                d.getType(),
                d.getVisibility(),
                d.getTitle(),
                d.getDescription(),
                d.getAgeRating(),
                d.getPriceRub(),
                locked,
                d.getOwnerUserId()
        );
    }

    public void assertCanAccessDeck(long userId, long deckId) {
        boolean premium = userRepository.findById(userId).orElseThrow().isPremium();
        Deck deck = deckRepository.findById(deckId).orElseThrow();

        if (!deck.isPublished()) {
            throw new IllegalArgumentException("Deck is not published");
        }

        // свои колоды доступны всегда
        if (Objects.equals(deck.getOwnerUserId(), userId)) return;

        // бесплатные доступны всем
        if (deck.getPriceRub() == 0) return;

        if (premium) {
            // премиум: PAID доступны, USER публичные доступны
            if ("PAID".equals(deck.getType())) return;
            if ("USER".equals(deck.getType()) && "PUBLIC".equals(deck.getVisibility())) return;
        } else {
            // без премиума: только купленные платные колоды
            boolean has = purchaseRepository.hasDeck(userId, deckId);
            if (has) return;
        }
        throw new IllegalArgumentException("Deck not accessible");
    }
}

File: src/main/kotlin/ru/trukhmanov/twochairsbackend/service/DeckService.kt
package ru.trukhmanov.twochairsbackend.service

import org.springframework.stereotype.Service
import ru.trukhmanov.twochairsbackend.dto.game.deck.DeckDto
import ru.trukhmanov.twochairsbackend.entity.Deck
import ru.trukhmanov.twochairsbackend.repository.UserPurchaseRepository
import ru.trukhmanov.twochairsbackend.repository.UserRepository
import ru.trukhmanov.twochairsbackend.repository.game.DeckRepository

@Service
class DeckService(
    private val deckRepository: DeckRepository,
    private val userRepository: UserRepository,
    private val purchaseRepository: UserPurchaseRepository
) {

    fun getAccessibleDecks(userId: Long): List<DeckDto> {
        val premium = userRepository.findById(userId).orElseThrow().premium

        val decks = if (premium) {
            deckRepository.findAccessibleForPremium(userId)
        } else {
            deckRepository.findAccessibleForNonPremium(userId)
        }

        return decks.map { toDto(it, false) }
    }

    fun getStoreDecks(userId: Long): List<DeckDto> {
        val premium = userRepository.findById(userId).orElseThrow().premium
        val decks = deckRepository.findStorePaidDecks()

        return decks.map { deck ->
            val locked = if (premium) {
                false
            } else {
                purchaseRepository.hasDeck(userId, requireNotNull(deck.id))
            }.not()
            toDto(deck, locked)
        }
    }

    private fun toDto(deck: Deck, locked: Boolean): DeckDto {
        return DeckDto(
            id = requireNotNull(deck.id),
            type = deck.type,
            visibility = deck.visibility,
            title = deck.title,
            description = deck.description,
            ageRating = deck.ageRating,
            priceRub = deck.priceRub,
            locked = locked,
            ownerUserId = deck.ownerUserId
        )
    }

    fun assertCanAccessDeck(userId: Long, deckId: Long) {
        val user = userRepository.findById(userId).orElseThrow()
        val deck = deckRepository.findById(deckId).orElseThrow()

        if (deck.ownerUserId == userId) {
            return
        }

        if (!deck.published) {
            throw IllegalArgumentException("Deck is not published")
        }

        val premium = user.premium

        if (deck.type == "DEFAULT") {
            return
        }

        if (deck.type == "USER" && !premium) {
            throw IllegalArgumentException("Deck not accessible")
        }

        if (premium) {
            if (deck.type == "PAID") return
            if (deck.type == "USER" && deck.visibility == "PUBLIC") return
            throw IllegalArgumentException("Deck not accessible")
        }

        if (deck.type == "PAID") {
            if (purchaseRepository.hasDeck(userId, deckId)) return
            throw IllegalArgumentException("Deck not accessible")
        }

        throw IllegalArgumentException("Deck not accessible")
    }
}
